crates/block-producer/src/mempool/mod.rs

```diff
@@ -0,0 +1,269 @@
+        // as part of a prior failure. Early exit to prevent logging these no-ops.
+        if removed_batches.is_empty() {
+            return;
+        }
+
+        let batches = removed_batches.iter().map(|(b, _)| *b).collect::<Vec<_>>();
+        let transactions = removed_batches.into_iter().flat_map(|(_, tx)| tx.into_iter()).collect();
+
+        self.transactions.requeue_transactions(transactions);
+
+        tracing::warn!(%batch, descendents=?batches, "Batch failed, dropping all inflight descendent batches, impacted transactions are back in queue.");
+    }
+
+    /// Marks a batch as proven if it exists.
+    pub fn batch_proved(&mut self, batch_id: BatchJobId) {
+        self.batches.mark_proven(batch_id);
+    }
+
+    /// Select at most `count` batches which are ready to be placed into the next block.
+    ///
+    /// May return an empty batch set if no batches are ready.
+    pub fn select_block(&mut self, count: usize) -> (BlockNumber, Vec<BatchJobId>) {
+        // TODO: should return actual batch transaction data as well.
+
+        let mut batches = Vec::with_capacity(count);
+        for _ in 0..count {
+            let Some((batch_id, _)) = self.batches.pop_for_blocking() else {
+                break;
+            };
+
+            batches.push(batch_id);
+
+            // Unlike `select_batch` we don't need to track inter-block depedencies as this
+            // relationship is inherently sequential.
+        }
+
+        let block_number = self.next_inflight_block;
+        self.next_inflight_block.increment();
+        self.block_pool.insert(block_number, batches.clone());
+
+        (block_number, batches)
+    }
+
+    /// Notify the pool that the block was succesfully completed.
+    ///
+    /// Panics if blocks are completed out-of-order. todo: might be a better way, but this is pretty
+    /// unrecoverable..
+    pub fn block_completed(&mut self, block_number: BlockNumber) {
+        assert_eq!(block_number, self.completed_blocks, "Blocks must be submitted sequentially");
+
+        // Update book keeping by removing the inflight data that just became stale.
+        self.completed_blocks.increment();
+
+        let Some(stale_block) = self.stale_block() else {
+            return;
+        };
+
+        let stale_batches = self.block_pool.remove(&stale_block).expect("Block should be in graph");
+
+        let stale_transations = self.batches.remove_stale(stale_batches);
+        self.transactions.remove_stale(stale_transations);
+    }
+
+    pub fn block_failed(&mut self, block: BlockNumber) {
+        // TBD.. not quite sure what to do here yet. Presumably the caller has already retried this
+        // block so the block is just inherently broken.
+        //
+        // Given lack of information at this stage we should probably just abort the node?
+        // In the future we might improve the situation with more fine-grained failure reasons.
+    }
```

Comment:
Right now I don't we can actually recover from this though. Current options as I see them:

1. Panic
2. Requeue the transactions and hope for a different set of batches (we could introduce randomness to help here)
3. Requeue batches 
4. Drop all transactions from the failed block

Reply:
I think in the short term, option 4 seems like a good one. This would allow dropping the "poisoned" transaction which makes the block production fail.

In the longer term, I think option 2 would be better and here we can keep a count of how many times a transaction has been re-queued. If it was re-queued more times than some limit (e.g., 3), we would drop it then. The idea is to reduce the number of transactions that we actually drop as presumably not all transactions in the failed block were bad.

Reply:
Makes sense. Ideally for (2) we would have more fine grained failure conditions which might pinpoint the problematic transaction(s).